keps/sig-instrumentation/20200415-cardinality-enforcement.md

```diff
@@ -0,0 +1,276 @@
+
+Even if we did have such a mechanism, I don't believe it would be a sufficient solution since:
+
+1. metrics changes can be seemingly innocous, even to the most diligent of users (this means it would be hard to catch at review)
+2. these cardinality issues exist latently; in other words, they're all over the place in Kubernetes already (so even if we could prevent 100% of these occurrence from **now on**, that wouldn't guarantee Kubernetes is free of these classes of issues).
+
+
+## Motivation
+
+TLDR; Having metrics turn into memory leaks sucks and it sucks even more when we can't fix these issues without re-releasing the entire Kubernetes binary.
+
+**Q:** *How have we approached these issues historically?*
+
+**A:** __Definitely not in a consistent manner.__
+
+Sometimes, a [metric label dimension is intended to be bound to some known sets of values but coding mistakes cause IDs to be thrown in as a label value](https://github.com/kubernetes/kubernetes/issues/53485).
+
+In anther case, [we opt to delete the entire metric](https://github.com/kubernetes/kubernetes/pull/74636) because it basically can't be used in a meaningfully correct way.
+
+Recently we've opted to both (1) [wholesale delete a metric label](https://github.com/kubernetes/kubernetes/pull/87669) and (2) [retroactively introduce a set of acceptable values for a metric label](https://github.com/kubernetes/kubernetes/pull/87913).
+
+Fixing these issues is a currently a manual process, both laborious and time-consuming.
+
+We don't have a standard prescription for resolving this class of issue. This is especially bad when you consider that this class of issue is so totally predictable (in general).
+
+### Goals
+
+This KEP proposes a possible solution for this, with the primary goal being to enable binding a metric dimension to a known set of values in such a way that is not coupled to code releases of Kubernetes.
+
+### Non-Goals
+
+We will expose the machinery and tools to bind a metric's labels to a discrete set of values.
+
+It is *not a goal* to implement and plumb this solution for each Kubernetes component (there are many SIGs and a number of verticals, which may have their own preferred way of doing things). As such it will be up to component owners to leverage this functionality that we provide, by feeding configuration data through whatever mechanism deemed appropriate (i.e. command line flags or reading from a file).
+
+## Proposal
+
+TLDR; we want to be able to *dynamically configure* a whitelist of label values for a metric.
+
+By *dynamically configure*, we mean configure a whitelist *at runtime* rather than during build/compile step (this is so ugh).
+
+And by *at runtime*, we mean, more specifically, during the boot sequence for a Kubernetes component (and we mean daemons here, not cli tools like kubectl).
```

Comment:
I get the value of being able to do this at runtime but... I know we already have plans to be able to dynamically disable metrics, I wonder if that plus compile time bounds on label length/values would be sufficient.

The reason I don't super like this is that it's not secure by default. Operators have to know what label configs to opt into for a given release, there isn't just a secure standard applied across the board, and the configurability makes it easy to override without necessarily understanding the risks involved ("I have tons of memory, why would I worry about this?").

Reply:
I like both label ideas :) 

RE the other point: There are some metrics we really can't turn off (i.e. SLO level metrics, like apiserver_request_total and the latency one) without creating a situation where you actually have very little visibility on what is happening in your cluster. So while disabling metrics is helpful for most cases, in special case situations (like the previous sec vuln), we need to be able to modify specific dimensions while maintaining the metric. 

Why is this not secure by default? Kubernetes is exactly as secure as it would be without this feature. This feature does not make anything less